util/src/test/java/io/zeebe/util/logging/StackdriverLayoutTest.java

```diff
@@ -0,0 +1,384 @@
+    try {
+      source.close();
+    } catch (final IOException e) {
+      LOGGER.error("Failed to close source input stream", e);
+    }
+
+    try {
+      sink.close();
+    } catch (final IOException e) {
+      LOGGER.error("Failed to close sink output stream", e);
+    }
+
+    logger.removeAppender(appender);
+  }
+
+  @Test
+  public void shouldWriteTraceMessage() throws IOException {
+    // given
+    logger.setLevel(Level.TRACE);
+
+    // when
+    logger.trace("Trace message");
+
+    // then
+    final var jsonMap = readLoggedEvent();
+    softly.assertThat(jsonMap).containsEntry("severity", Severity.DEBUG.name());
+  }
+
+  @Test
+  public void shouldWriteDebugMessages() throws IOException {
+    // given
+    logger.setLevel(Level.DEBUG);
+
+    // when
+    logger.debug("Debug message");
+
+    // then
+    final var jsonMap = readLoggedEvent();
+    softly.assertThat(jsonMap).containsEntry("severity", Severity.DEBUG.name());
+  }
+
+  @Test
+  public void shouldWriteInfoMessage() throws IOException {
+    // given
+    logger.setLevel(Level.INFO);
+
+    // when
+    logger.info("Info message");
+
+    // then
+    final var jsonMap = readLoggedEvent();
+    softly.assertThat(jsonMap).containsEntry("severity", Severity.INFO.name());
+  }
+
+  @Test
+  public void shouldWriteWarningMessage() throws IOException {
+    // given
+    logger.setLevel(Level.WARN);
+
+    // when
+    logger.warn("Info message");
+
+    // then
+    final var jsonMap = readLoggedEvent();
+    softly.assertThat(jsonMap).containsEntry("severity", Severity.WARNING.name());
+  }
+
+  @Test
+  public void shouldWriteErrorMessageWithoutException() throws IOException {
+    // given
+    logger.setLevel(Level.ERROR);
+
+    // when
+    logger.error("Error message {}", 1);
+    final var stackTrace = new IllegalStateException("").getStackTrace();
+    final var source = stackTrace[0];
+
+    // then
+    final var jsonMap = readLoggedEvent();
+    softly
+        .assertThat(jsonMap)
+        .containsEntry("severity", Severity.ERROR.name())
+        .containsEntry("message", "Error message 1")
+        .containsEntry("@type", StackdriverLogEntry.ERROR_REPORT_TYPE)
+        .hasEntrySatisfying(
+            "context",
+            context ->
+                softly
+                    .assertThat(context)
+                    .asInstanceOf(InstanceOfAssertFactories.MAP)
+                    .hasEntrySatisfying(
+                        StackdriverLogEntryBuilder.ERROR_REPORT_LOCATION_CONTEXT_KEY,
+                        reportLocation ->
+                            softly
+                                .assertThat(reportLocation)
+                                .asInstanceOf(InstanceOfAssertFactories.MAP)
+                                .containsEntry("file", source.getFileName())
+                                .containsEntry("function", source.getMethodName())
+                                .containsEntry("line", source.getLineNumber() - 1)))
```

Comment:
The source location is the file, method, and line at which the logging statement was called (e.g. `logger.info`). Since I don't want to hard code the line (so when you add code above you need to update this test), but I also don't know how to get the stack trace other than through an exception (:see_no_evil:), I ended up creating an exception right after the logging statement, and using its stack trace to figure out the line of my logging statement :sweat_smile: 

If you know how to get the line of the logging statement programmatically I'm all ears :slightly_smiling_face:

Reply:
I had to use my duckduckgo-foo here, but I found this: https://stackoverflow.com/a/25552778 `Thread.currentThread().getStackTrace()`.